embassy-net-adin1110/src/lib.rs

```diff
@@ -0,0 +1,1249 @@
+        spi_hdr.set_write(true);
+        spi_hdr.set_addr(reg);
+        let _ = tx_buf.extend_from_slice(spi_hdr.0.to_be_bytes().as_slice());
+
+        if self.crc {
+            // Add CRC for header data
+            let _ = tx_buf.push(crc8(&tx_buf));
+        }
+
+        let val = value.to_be_bytes();
+        let _ = tx_buf.extend_from_slice(val.as_slice());
+
+        if self.crc {
+            // Add CRC for header data
+            let _ = tx_buf.push(crc8(val.as_slice()));
+        }
+
+        #[cfg(feature = "defmt")]
+        defmt::trace!("REG Write {} = {:08x} SPI {:02x}", reg, value, &tx_buf);
+
+        self.spi.write(&tx_buf).await.map_err(AdinError::Spi)
+    }
+
+    // helper function for write to MDIO_ACC register and wait for ready!
+    async fn write_mdio_acc_reg(&mut self, mdio_acc_val: u32) -> AEResult<u32, SpiE> {
+        self.write_reg(sr::MDIO_ACC, mdio_acc_val).await?;
+
+        // TODO: Add proper timeout!
+        for _ in 0..100_000 {
+            let val = self.read_reg(sr::MDIO_ACC).await?;
+            if val & 0x8000_0000 != 0 {
+                return Ok(val);
+            }
+        }
+
+        Err(AdinError::MDIO_ACC_TIMEOUT)
+    }
+
+    pub async fn read_fifo(&mut self, packet: &mut [u8]) -> AEResult<usize, SpiE> {
+        let mut tx_buf = Vec::<u8, 16>::new();
+
+        // Size of the frame, also includes the appednded header.
+        let packet_size = self.read_reg(sr::RX_FSIZE).await?;
+
+        // Packet read of write to the MAC packet buffer must be a multipul of 4!
+        let read_size = size_align_u32(packet_size);
+
+        if packet_size < u32::try_from(FRAME_HEADER_LEN + FEC_LEN).unwrap()
+            || read_size > u32::try_from(packet.len()).unwrap()
+        {
+            return Err(AdinError::PACKET_TOO_BIG);
+        }
+
+        let mut spi_hdr = SpiHeader(0);
+        spi_hdr.set_control(true);
+        spi_hdr.set_addr(sr::RX);
+        let _ = tx_buf.extend_from_slice(spi_hdr.0.to_be_bytes().as_slice());
+
+        if self.crc {
+            // Add CRC for header data
+            let _ = tx_buf.push(crc8(&tx_buf));
+        }
+
+        // Turn around byte, TODO: Unknown that this is.
+        let _ = tx_buf.push(TURN_AROUND_BYTE);
+
+        let spi_packet = &mut packet[0..read_size as usize];
+
+        assert_eq!(spi_packet.len() & 0x03, 0x00);
+
+        let mut pkt_header = [0, 0];
+
+        let mut spi_op = [
+            Operation::Write(&tx_buf),
+            Operation::Read(&mut pkt_header),
+            Operation::Read(spi_packet),
+        ];
+
+        self.spi.transaction(&mut spi_op).await.map_err(AdinError::Spi)?;
+
+        Ok(packet_size as usize)
+    }
+
+    pub async fn write_fifo(&mut self, frame: &[u8]) -> AEResult<(), SpiE> {
+        let header_len = self.header_write_len();
+        // if packet.len() < header_len {
+        //     return Err(AdinError::PACKET_TOO_SMALL);
+        // }
+
+        let mut packet = Packet::new();
```

Comment:
this uses 1500 bytes of memory. you can reduce this if you allocate space for only the header, and then do an SPI transaction which writes the header, and then the frame straight from the `frame` slice.

Reply:
I see what you are trying to say.

But I have this issue, the question what is the best option.

Create one buffer/vector, put all the data in it and start one SPI transfer.
This is the linux kernel way: https://elixir.bootlin.com/linux/latest/source/drivers/net/ethernet/adi/adin1110.c#L416
Data is <SPI HEADER 2b><OPTIONAL CRC 1b><PORT BYTES 2b><FRAME xb><OPTIONAL PADDING to 64b><FCS 2b><MAKE THE FRAME EVEN BYTES 1b>

Or split this up in 3 sections
HEAD: <SPI HEADER 2b><OPTIONAL CRC 1b><PORT BYTES 2b>
FRAME: <FRAME xb>
TRAIL: <OPTIONAL PADDING to 64b><FCS 2b><MAKE THE FRAME EVEN BYTES 1b>
and do a `self.transaction(&mut [Operation::Write(HEAD), Operation::Write(FRAME), Operation::Write(TRAIL))])`
I think that this also translates in to 3 expensive DMA SPI s transactions.

For reading the chipś fifo, it is a similar issue.
But the chip supports that spi write and spi read is one transaction.
I could use transfer_in_place() for this operation.

What do you prefer?

Reply:
The problem with using a single buffer is it is quite big, it needs space for the MTU and headers, which is a bit more than 1.5kb. If your microcontroller has 128kb, 256kb of RAM, saving 1.5kb is a big deal.

Yes, starting 3 separate DMA transfers has some overhead. It is probably smaller than copying the packet from one buffer to the other.

Reply:
Thanks for the feedback.